### PR TITLE
JENKINS-42763: Make override of HOME configurable

### PR DESCRIPTION
**Pull request by TobiX**
This is basically a "I'm so annoyed by this, I actually wrote a fix" PR :smiling_imp:

This should fix [JENKINS-42763](https://issues.jenkins-ci.org/browse/JENKINS-42763) by making that
"feature" configurable. Since I consider NOT overwriting the HOME environment variable the correct behaviour, I tried to implement the following features:

- New clouds are created with the option disabled
- When updating old data, the option is enabled, so behaviour is unchanged for existing clouds...

There seems to be one minor behaviour change in a corner case, see test cases.

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 1]
```diff
@@ -286,12 +286,14 @@ public Pod build() {
                 	env.put("http_proxy", httpProxy);
                 }
             }
-        }
 
-        // Running on OpenShift Enterprise, security concerns force use of arbitrary user ID
-        // As a result, container is running without a home set for user, resulting into using `/` for some tools,
-        // and `?` for java build tools. So we force HOME to a safe location.
-        env.put("HOME", workingDir);
+            if (slave.getKubernetesCloud().isEnsureHomeIsValid()) {
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: This does not seem to allow control from the `podTemplate` step.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 2]
```diff
@@ -624,6 +635,7 @@ public boolean equals(Object o) {
         KubernetesCloud that = (KubernetesCloud) o;
         return skipTlsVerify == that.skipTlsVerify &&
                 addMasterProxyEnvVars == that.addMasterProxyEnvVars &&
+                ensureHomeIsValid == that.ensureHomeIsValid &&
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]:
```
[ERROR] Suspicious comparison of Boolean references in org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud.equals(Object) [org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud] At KubernetesCloud.java:[line 636] RC_REF_COMPARISON_BAD_PRACTICE_BOOLEAN
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 1 above)
Comment (TobiX) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: Drat. I shouldn't make pull requests right before going to bed 😴

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 2 above)
Comment (TobiX) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: I changed the type of that from `boolean` to `Boolean` quite late, so I missed that... Will take a look.

**Comment by TobiX** [comment]
> Is the problem with `?` for home acc. to Java really specific to OpenShift? I know I have seen this in other contexts, but not sure about the `kubernetes` plugin specifically.

Now that I think about it: It is probably specific to your container setup. For my local setup, it seems the `USER` of the jnlp container is the one everything is run under. If you make sure that UID exists on all containers you are using, you have no problems... If the UID is totally random (like it seems to be the case for OpenShift), you are out of luck regardless and overriding the HOME directory is a "sensible" thing...

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 1 above)
Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: I mean, it is not a mistake, just a limitation which we may decide is good enough for now. Basically means that when defining build environments in `Jenkinsfile` rather than global settings, you do not have the opportunity to decide on a case-by-case basis whether or not to use the `$HOME` fix.

**Comment by Vlatombe** [comment]
Closed in favor of #531
